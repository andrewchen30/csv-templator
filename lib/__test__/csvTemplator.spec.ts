import CSVTemplator from '@/index';

describe('CSV Templator', () => {
  describe('parse schema', () => {
    const successCases: {
      desc: string;
      template: string;
    }[] = [
      {
        desc: 'should parse data cell (Horizontal)',
        template: `
          | 'aaa' |
          | 'bbb' |
          | 'ccc' |
        `,
      },
      {
        desc: 'should parse data cell (Vertical)',
        template: `
          | 'aaa' | 'bbb' | 'ccc' |
        `,
      },
      {
        desc: 'should parse data cell (Mixed)',
        template: `
          | 'aaa' | 'bbb' |
          | 'ccc' | 'ddd' |
          | 'eee' | 'fff' |
        `,
      },
      {
        desc: 'should parse for-col',
        template: `
          | // for-col item in list | 
          | item.name | 
          | item.name |
          | item.name |
        `,
      },
      {
        desc: 'should parse for-row',
        template: `
          | // for-row item in list | item.name | item.name | item.name |
        `,
      },
      {
        desc: 'should parse extend-col and extend-row',
        template: `

        | | // for-col item in list |
        | | item.name |
        | | item.name |
        | | item.name |
        | // for-row item2 in list | item2[item.name] |
      `,
      },
    ];

    it.each(successCases)('$desc', ({ template }) => {
      const templator = new CSVTemplator();
      templator.useTemplate(template);

      expect(templator._schema).toMatchSnapshot('_schema');
      expect(templator._logicColIndexes).toMatchSnapshot('_logicColIndexes');
      expect(templator._logicRowIndexes).toMatchSnapshot('_logicRowIndexes');
    });
  });

  describe('prepare render data', () => {
    const successCases: {
      desc: string;
      template: string;
      data: any;
    }[] = [
      {
        desc: 'should successfully prepare data for render',
        template: `
        | | "name" | "age"
        | // for-row user in users | user.name | user.age |
      `,
        data: {
          users: [
            { name: 'Andrew', age: 30 },
            { name: 'Joanne', age: 28 },
            { name: 'Frank', age: 22 },
          ],
        },
      },
    ];

    it.each(successCases)('$desc', ({ template, data }) => {
      const templator = new CSVTemplator();
      templator.useTemplate(template);
      templator.render(data);

      expect(templator._rawValue).toMatchSnapshot('_rawValue');
    });
  });
});
