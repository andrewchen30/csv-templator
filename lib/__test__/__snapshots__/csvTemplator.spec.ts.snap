// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSV Templator parse schema should parse data cell (Horizontal): _logicColIndexes 1`] = `Set {}`;

exports[`CSV Templator parse schema should parse data cell (Horizontal): _logicRowIndexes 1`] = `Set {}`;

exports[`CSV Templator parse schema should parse data cell (Horizontal): _schema 1`] = `
Array [
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 0,
      },
      "eval": "'aaa'",
      "renderPosition": Object {
        "col": 0,
        "row": 0,
      },
      "type": 1,
    },
  ],
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 1,
      },
      "eval": "'bbb'",
      "renderPosition": Object {
        "col": 0,
        "row": 1,
      },
      "type": 1,
    },
  ],
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 2,
      },
      "eval": "'ccc'",
      "renderPosition": Object {
        "col": 0,
        "row": 2,
      },
      "type": 1,
    },
  ],
]
`;

exports[`CSV Templator parse schema should parse data cell (Mixed): _logicColIndexes 1`] = `Set {}`;

exports[`CSV Templator parse schema should parse data cell (Mixed): _logicRowIndexes 1`] = `Set {}`;

exports[`CSV Templator parse schema should parse data cell (Mixed): _schema 1`] = `
Array [
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 0,
      },
      "eval": "'aaa'",
      "renderPosition": Object {
        "col": 0,
        "row": 0,
      },
      "type": 1,
    },
    Object {
      "_positionInTemplate": Object {
        "col": 1,
        "row": 0,
      },
      "eval": "'bbb'",
      "renderPosition": Object {
        "col": 1,
        "row": 0,
      },
      "type": 1,
    },
  ],
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 1,
      },
      "eval": "'ccc'",
      "renderPosition": Object {
        "col": 0,
        "row": 1,
      },
      "type": 1,
    },
    Object {
      "_positionInTemplate": Object {
        "col": 1,
        "row": 1,
      },
      "eval": "'ddd'",
      "renderPosition": Object {
        "col": 1,
        "row": 1,
      },
      "type": 1,
    },
  ],
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 2,
      },
      "eval": "'eee'",
      "renderPosition": Object {
        "col": 0,
        "row": 2,
      },
      "type": 1,
    },
    Object {
      "_positionInTemplate": Object {
        "col": 1,
        "row": 2,
      },
      "eval": "'fff'",
      "renderPosition": Object {
        "col": 1,
        "row": 2,
      },
      "type": 1,
    },
  ],
]
`;

exports[`CSV Templator parse schema should parse data cell (Vertical): _logicColIndexes 1`] = `Set {}`;

exports[`CSV Templator parse schema should parse data cell (Vertical): _logicRowIndexes 1`] = `Set {}`;

exports[`CSV Templator parse schema should parse data cell (Vertical): _schema 1`] = `
Array [
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 0,
      },
      "eval": "'aaa'",
      "renderPosition": Object {
        "col": 0,
        "row": 0,
      },
      "type": 1,
    },
    Object {
      "_positionInTemplate": Object {
        "col": 1,
        "row": 0,
      },
      "eval": "'bbb'",
      "renderPosition": Object {
        "col": 1,
        "row": 0,
      },
      "type": 1,
    },
    Object {
      "_positionInTemplate": Object {
        "col": 2,
        "row": 0,
      },
      "eval": "'ccc'",
      "renderPosition": Object {
        "col": 2,
        "row": 0,
      },
      "type": 1,
    },
  ],
]
`;

exports[`CSV Templator parse schema should parse extend-col and extend-row: _logicColIndexes 1`] = `
Set {
  0,
}
`;

exports[`CSV Templator parse schema should parse extend-col and extend-row: _logicRowIndexes 1`] = `
Set {
  0,
}
`;

exports[`CSV Templator parse schema should parse extend-col and extend-row: _schema 1`] = `
Array [
  Array [
    null,
    Object {
      "_positionInTemplate": Object {
        "col": 1,
        "row": 0,
      },
      "logicType": 0,
      "loopArgs": Array [
        "item",
      ],
      "targetArray": "list",
      "type": 0,
    },
  ],
  Array [
    null,
    Object {
      "_positionInTemplate": Object {
        "col": 1,
        "row": 1,
      },
      "eval": "item.name",
      "renderPosition": Object {
        "col": 1,
        "row": 1,
      },
      "type": 1,
    },
  ],
  Array [
    null,
    Object {
      "_positionInTemplate": Object {
        "col": 1,
        "row": 2,
      },
      "eval": "item.name",
      "renderPosition": Object {
        "col": 1,
        "row": 2,
      },
      "type": 1,
    },
  ],
  Array [
    null,
    Object {
      "_positionInTemplate": Object {
        "col": 1,
        "row": 3,
      },
      "eval": "item.name",
      "renderPosition": Object {
        "col": 1,
        "row": 3,
      },
      "type": 1,
    },
  ],
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 4,
      },
      "logicType": 1,
      "loopArgs": Array [
        "item2",
      ],
      "targetArray": "list",
      "type": 0,
    },
    Object {
      "_positionInTemplate": Object {
        "col": 1,
        "row": 4,
      },
      "eval": "item2[item.name]",
      "renderPosition": Object {
        "col": 1,
        "row": 4,
      },
      "type": 1,
    },
  ],
]
`;

exports[`CSV Templator parse schema should parse for-col: _logicColIndexes 1`] = `Set {}`;

exports[`CSV Templator parse schema should parse for-col: _logicRowIndexes 1`] = `
Set {
  0,
}
`;

exports[`CSV Templator parse schema should parse for-col: _schema 1`] = `
Array [
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 0,
      },
      "logicType": 0,
      "loopArgs": Array [
        "item",
      ],
      "targetArray": "list",
      "type": 0,
    },
    null,
  ],
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 1,
      },
      "eval": "item.name",
      "renderPosition": Object {
        "col": 0,
        "row": 1,
      },
      "type": 1,
    },
    null,
  ],
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 2,
      },
      "eval": "item.name",
      "renderPosition": Object {
        "col": 0,
        "row": 2,
      },
      "type": 1,
    },
    null,
  ],
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 3,
      },
      "eval": "item.name",
      "renderPosition": Object {
        "col": 0,
        "row": 3,
      },
      "type": 1,
    },
    null,
  ],
]
`;

exports[`CSV Templator parse schema should parse for-row: _logicColIndexes 1`] = `
Set {
  0,
}
`;

exports[`CSV Templator parse schema should parse for-row: _logicRowIndexes 1`] = `Set {}`;

exports[`CSV Templator parse schema should parse for-row: _schema 1`] = `
Array [
  Array [
    Object {
      "_positionInTemplate": Object {
        "col": 0,
        "row": 0,
      },
      "logicType": 1,
      "loopArgs": Array [
        "item",
      ],
      "targetArray": "list",
      "type": 0,
    },
    Object {
      "_positionInTemplate": Object {
        "col": 1,
        "row": 0,
      },
      "eval": "item.name",
      "renderPosition": Object {
        "col": 1,
        "row": 0,
      },
      "type": 1,
    },
    Object {
      "_positionInTemplate": Object {
        "col": 2,
        "row": 0,
      },
      "eval": "item.name",
      "renderPosition": Object {
        "col": 2,
        "row": 0,
      },
      "type": 1,
    },
    Object {
      "_positionInTemplate": Object {
        "col": 3,
        "row": 0,
      },
      "eval": "item.name",
      "renderPosition": Object {
        "col": 3,
        "row": 0,
      },
      "type": 1,
    },
  ],
]
`;
